sdl_android/src/main/java/com/smartdevicelink/managers/SdlManager.java

```diff
@@ -129,94 +127,137 @@ public void onError(String info, Exception e){
-				if(managerListener != null){
-					managerListener.onStart();
-				}
+			checkState();
+		}
+	};
 
-				// Set the app icon
-				if (SdlManager.this.appIcon != null && SdlManager.this.appIcon.getName() != null) {
-					if (!fileManager.hasUploadedFile(SdlManager.this.appIcon)) {
-						fileManager.uploadArtwork(SdlManager.this.appIcon, new CompletionListener() {
-							@Override
-							public void onComplete(boolean success) {
-								if (success) {
-									SetAppIcon msg = new SetAppIcon(SdlManager.this.appIcon.getName());
-									_internalInterface.sendRPCRequest(msg);
-								}
-							}
-						});
-					} else {
-						SetAppIcon msg = new SetAppIcon(SdlManager.this.appIcon.getName());
-						_internalInterface.sendRPCRequest(msg);
+	void checkState() {
+		if (permissionManager != null && fileManager != null && screenManager != null && (!lockScreenConfig.isEnabled() || lockScreenManager != null)) {
+			if (permissionManager.getState() == BaseSubManager.READY && fileManager.getState() == BaseSubManager.READY && screenManager.getState() == BaseSubManager.READY && (!lockScreenConfig.isEnabled() || lockScreenManager.getState() == BaseSubManager.READY)) {
+				transitionToState(BaseSubManager.READY);
+				notifyDevListener(null);
+			} else if (permissionManager.getState() == BaseSubManager.ERROR && fileManager.getState() == BaseSubManager.ERROR && screenManager.getState() == BaseSubManager.ERROR && (!lockScreenConfig.isEnabled() || lockScreenManager.getState() == BaseSubManager.ERROR)) {
+				String info = "ERROR starting sdl manager, all sub managers are in error state";
+				DebugTool.logError(info);
+				transitionToState(BaseSubManager.ERROR);
+				notifyDevListener(info);
+			} else if (permissionManager.getState() == BaseSubManager.SETTING_UP || fileManager.getState() == BaseSubManager.SETTING_UP || screenManager.getState() == BaseSubManager.SETTING_UP || (lockScreenConfig.isEnabled() && lockScreenManager != null && lockScreenManager.getState() == BaseSubManager.SETTING_UP)) {
+				transitionToState(BaseSubManager.SETTING_UP);
+				// No need to notify developer here!
+			} else {
+				DebugTool.logWarning("LIMITED starting sdl manager, some sub managers are in error or limited state and the others finished setting up");
+				transitionToState(BaseSubManager.LIMITED);
+				notifyDevListener(null);
+			}
+		} else {
+			// We should never be here, but somehow one of the sub-sub managers is null
+			String info = "ERROR one of the sdl sub managers is null";
+			DebugTool.logError(info);
+			transitionToState(BaseSubManager.ERROR);
+			notifyDevListener(info);
+		}
+	}
+
+	private void notifyDevListener(String info) {
+		if (managerListener != null) {
+			if (state == BaseSubManager.ERROR){
+				managerListener.onError(info, null);
+			} else {
+				managerListener.onStart();
+			}
+		}
+
+		// Set the app icon
+		if (state == BaseSubManager.READY || state == BaseSubManager.LIMITED) {
+			setAppIcon();
+		}
```

Comment:
This logic seems a little out of place/out of scope of the method that it is in. This should probably be in the `checkState` method or the `transitionToState` method.

Reply:
Or create a new method `onReady` in the manager that can handle anything that needs to be handled during the transition to READY.

Reply:
Makes sense. I will create `onReady()` method to handle that logic.